File: backend/routers/videos.py
from fastapi import APIRouter, HTTPException, Query
from ..models.video import VideoListResponse, VideoDetail, VideoItem
from ..services.scraper import scraper_service
from ..services.video_cache import video_cache_service
from ..services.proxy import proxy_service
from ..config import settings
import asyncio

router = APIRouter(prefix="/api/videos", tags=["videos"])

_total_pages_cache = {"total_pages": 1}
_precache_queue = set()  # 正在预缓存的视频ID

# 列表缓存有效期：3小时
LIST_CACHE_MAX_AGE = 3 * 60 * 60


async def _download_thumbnails(videos):
    """后台下载视频封面图"""
    for video in videos:
        if video.thumbnail:
            await video_cache_service.download_thumbnail(video.id, video.thumbnail)


async def _precache_video(video_id: str):
    """预缓存单个视频"""
    # 检查是否已缓存或正在下载
    if video_cache_service.is_cached(video_id):
        return
    if video_cache_service.is_downloading(video_id):
        return
    if video_id in _precache_queue:
        return

    _precache_queue.add(video_id)
    try:
        video_url = f"{settings.TARGET_BASE_URL}/view_video.php?viewkey={video_id}"
        detail = await scraper_service.get_video_detail_in_new_tab(video_url)

        if not detail or not detail.m3u8_url:
            print(f"[预缓存] 跳过 {video_id}: 无法获取视频链接")
            return

        # 再次检查（可能在获取详情期间用户已开始播放）
        if video_cache_service.is_cached(video_id) or video_cache_service.is_downloading(video_id):
            return

        # 启动缓存下载
        video_src = detail.m3u8_url
        is_mp4 = ".mp4" in video_src.lower() or ".m3u8" not in video_src.lower()

        if is_mp4:
            await video_cache_service.start_mp4_cache_download(video_id, video_src, detail)
        else:
            # 获取 m3u8 内容
            session = await proxy_service.get_session()
            async with session.get(video_src) as resp:
                original_m3u8 = await resp.text()
            await video_cache_service.start_cache_download(video_id, video_src, original_m3u8, detail)

        print(f"[预缓存] 已启动: {video_id}")

    except Exception as e:
        print(f"[预缓存] 失败 {video_id}: {e}")
    finally:
        _precache_queue.discard(video_id)


async def _precache_videos(videos):
    """后台预缓存视频列表"""
    # 使用信号量限制并发数
    semaphore = asyncio.Semaphore(settings.PRECACHE_CONCURRENT)

    async def precache_with_limit(video):
        async with semaphore:
            await _precache_video(video.id)

    # 并发预缓存
    tasks = [precache_with_limit(v) for v in videos]
    await asyncio.gather(*tasks, return_exceptions=True)


@router.get("", response_model=VideoListResponse)
async def get_video_list(page: int = Query(1, ge=1, description="页码")):
    """获取视频列表"""
    # 优先使用有效期内的缓存（3小时）
    if settings.VIDEO_CACHE_ENABLED:
        fresh_cache = await video_cache_service.get_cached_list(page, max_age=LIST_CACHE_MAX_AGE)
        if fresh_cache:
            videos = [VideoItem(**v) for v in fresh_cache.get("videos", [])]
            response = VideoListResponse(
                videos=videos,
                total=fresh_cache.get("total", len(videos)),
                page=fresh_cache.get("page", page),
                total_pages=fresh_cache.get("total_pages", 1)
            )
            if response.total_pages > 1:
                _total_pages_cache["total_pages"] = response.total_pages
            return response

    # 缓存过期或不存在，尝试从网站获取
    result = None
    fetch_error = None
    try:
        result = await scraper_service.get_video_list(page_num=page)
    except Exception as e:
        fetch_error = e
        print(f"获取视频列表失败: {e}")

    # 获取成功且有数据
    if result and result.videos:
        # 更新总页数缓存
        if result.total_pages > 1:
            _total_pages_cache["total_pages"] = result.total_pages

        response = VideoListResponse(
            videos=result.videos,
            total=len(result.videos),
            page=page,
            total_pages=_total_pages_cache["total_pages"]
        )

        # 保存到文件缓存
        if settings.VIDEO_CACHE_ENABLED:
            await video_cache_service.save_list_cache(page, {
                "videos": [v.model_dump() for v in result.videos],
                "total": len(result.videos),
                "page": page,
                "total_pages": _total_pages_cache["total_pages"]
            })
            # 后台异步下载封面图
            asyncio.create_task(_download_thumbnails(result.videos))
            # 后台异步预缓存视频
            if settings.AUTO_PRECACHE:
                asyncio.create_task(_precache_videos(result.videos))

        return response

    # 获取失败或无数据，尝试使用过期的缓存作为兜底
    if settings.VIDEO_CACHE_ENABLED:
        file_cached = await video_cache_service.get_cached_list(page)  # 不检查时间
        if file_cached:
            videos = [VideoItem(**v) for v in file_cached.get("videos", [])]
            response = VideoListResponse(
                videos=videos,
                total=file_cached.get("total", len(videos)),
                page=file_cached.get("page", page),
                total_pages=file_cached.get("total_pages", 1)
            )
            # 更新总页数
            if response.total_pages > 1:
                _total_pages_cache["total_pages"] = response.total_pages
            print(f"[Cache] 使用过期缓存兜底: 第{page}页, {len(videos)}个视频")
            return response

    # 既无法获取也无缓存
    if fetch_error:
        raise HTTPException(status_code=500, detail=f"获取视频列表失败: {str(fetch_error)}")
    else:
        raise HTTPException(status_code=404, detail="暂无视频数据")


@router.get("/{video_id}", response_model=VideoDetail)
async def get_video_detail(video_id: str):
    """获取视频详情"""
    from ..services.video_cache import video_cache_service

    # 如果视频文件已缓存，优先使用持久化的详情缓存
    if video_cache_service.is_cached(video_id):
        cached_detail = await video_cache_service.get_cached_detail(video_id)
        if cached_detail:
            return cached_detail

    # 视频未缓存，每次都重新获取详情（不使用内存缓存）
    try:
        video_url = f"{settings.TARGET_BASE_URL}/view_video.php?viewkey={video_id}"
        detail = await scraper_service.get_video_detail(video_url)

        if not detail:
            raise HTTPException(status_code=404, detail="视频不存在")

        return detail

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取视频详情失败: {str(e)}")


@router.delete("/cache")
async def clear_cache():
    """清除缓存"""
    _total_pages_cache["total_pages"] = 1
    return {"message": "缓存已清除"}

File: backend-go/routers/videos.go
package routers

import (
	"backend-go/config"
	"backend-go/models"
	"backend-go/services"
	"log"
	"net/http"
	"strconv"
	"strings"
	"sync"

	"github.com/gin-gonic/gin"
)

var (
	totalPagesCache = struct {
		sync.RWMutex
		value int
	}{value: 1}

	precacheQueue = struct {
		sync.RWMutex
		set map[string]bool
	}{set: make(map[string]bool)}
)

// RegisterVideosRoutes 注册视频相关路由
func RegisterVideosRoutes(r *gin.RouterGroup) {
	videos := r.Group("/videos")
	{
		videos.GET("", getVideoList)
		videos.GET("/:video_id", getVideoDetail)
		videos.DELETE("/cache", clearVideoCache)
	}
}

// getVideoList 获取视频列表
func getVideoList(c *gin.Context) {
	page := 1
	if p := c.Query("page"); p != "" {
		if v, err := strconv.Atoi(p); err == nil && v > 0 {
			page = v
		}
	}

	cfg := config.Settings
	cacheService := services.GetVideoCacheService()
	scraperService := services.GetScraperService()

	// 优先使用有效期内的缓存
	if cfg.VideoCacheEnabled {
		freshCache, err := cacheService.GetCachedList(page, cfg.VideoListCacheTTL)
		if err == nil && freshCache != nil {
			videos := parseVideosFromCache(freshCache)
			total := getIntFromMap(freshCache, "total", len(videos))
			totalPages := getIntFromMap(freshCache, "total_pages", 1)

			if totalPages > 1 {
				totalPagesCache.Lock()
				totalPagesCache.value = totalPages
				totalPagesCache.Unlock()
			}

			c.JSON(http.StatusOK, models.VideoListResponse{
				Videos:     videos,
				Total:      total,
				Page:       page,
				TotalPages: totalPages,
			})
			return
		}
	}

	// 缓存过期或不存在，尝试从网站获取
	var result *services.VideoListResult
	var fetchError error

	result, fetchError = scraperService.GetVideoList(page)

	if fetchError != nil {
		log.Printf("获取视频列表失败: %v", fetchError)
	}

	// 获取成功且有数据
	if result != nil && len(result.Videos) > 0 {
		if result.TotalPages > 1 {
			totalPagesCache.Lock()
			totalPagesCache.value = result.TotalPages
			totalPagesCache.Unlock()
		}

		totalPagesCache.RLock()
		tp := totalPagesCache.value
		totalPagesCache.RUnlock()

		response := models.VideoListResponse{
			Videos:     result.Videos,
			Total:      len(result.Videos),
			Page:       page,
			TotalPages: tp,
		}

		// 保存到文件缓存
		if cfg.VideoCacheEnabled {
			videoMaps := make([]map[string]interface{}, len(result.Videos))
			for i, v := range result.Videos {
				videoMaps[i] = map[string]interface{}{
					"id":        v.ID,
					"title":     v.Title,
					"thumbnail": v.Thumbnail,
					"url":       v.URL,
					"duration":  v.Duration,
				}
			}

			cacheData := map[string]interface{}{
				"videos":      videoMaps,
				"total":       len(result.Videos),
				"page":        page,
				"total_pages": tp,
			}
			cacheService.SaveListCache(page, cacheData)

			// 后台异步下载封面图
			go downloadThumbnails(result.Videos)

			// 后台异步预缓存视频
			if cfg.AutoPrecache {
				go precacheVideos(result.Videos)
			}
		}

		c.JSON(http.StatusOK, response)
		return
	}

	// 获取失败或无数据，尝试使用过期的缓存作为兜底
	if cfg.VideoCacheEnabled {
		fileCached, err := cacheService.GetCachedList(page, 0) // 不检查时间
		if err == nil && fileCached != nil {
			videos := parseVideosFromCache(fileCached)
			total := getIntFromMap(fileCached, "total", len(videos))
			totalPages := getIntFromMap(fileCached, "total_pages", 1)

			if totalPages > 1 {
				totalPagesCache.Lock()
				totalPagesCache.value = totalPages
				totalPagesCache.Unlock()
			}

			log.Printf("[Cache] 使用过期缓存兜底: 第%d页, %d个视频", page, len(videos))
			c.JSON(http.StatusOK, models.VideoListResponse{
				Videos:     videos,
				Total:      total,
				Page:       page,
				TotalPages: totalPages,
			})
			return
		}
	}

	// 既无法获取也无缓存
	if fetchError != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{
			Detail: "获取视频列表失败: " + fetchError.Error(),
		})
		return
	}

	c.JSON(http.StatusNotFound, models.ErrorResponse{
		Detail: "暂无视频数据",
	})
}

// getVideoDetail 获取视频详情
func getVideoDetail(c *gin.Context) {
	videoID := c.Param("video_id")
	cacheService := services.GetVideoCacheService()
	scraperService := services.GetScraperService()
	cfg := config.Settings

	// 如果视频文件已缓存，优先使用持久化的详情缓存
	if cacheService.IsCached(videoID) {
		cachedDetail, err := cacheService.GetCachedDetail(videoID)
		if err == nil && cachedDetail != nil {
			c.JSON(http.StatusOK, cachedDetail)
			return
		}
	}

	// 视频未缓存，每次都重新获取详情（使用新标签页避免冲突）
	videoURL := cfg.TargetBaseURL + "/view_video.php?viewkey=" + videoID
	detail, err := scraperService.GetVideoDetailInNewTab(videoURL)

	if err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{
			Detail: "获取视频详情失败: " + err.Error(),
		})
		return
	}

	if detail == nil {
		c.JSON(http.StatusNotFound, models.ErrorResponse{
			Detail: "视频不存在",
		})
		return
	}

	c.JSON(http.StatusOK, detail)
}

// clearVideoCache 清除缓存
func clearVideoCache(c *gin.Context) {
	totalPagesCache.Lock()
	totalPagesCache.value = 1
	totalPagesCache.Unlock()

	c.JSON(http.StatusOK, gin.H{"message": "缓存已清除"})
}

// 辅助函数
func parseVideosFromCache(data map[string]interface{}) []models.VideoItem {
	var videos []models.VideoItem

	if videosData, ok := data["videos"].([]interface{}); ok {
		for _, v := range videosData {
			if vm, ok := v.(map[string]interface{}); ok {
				video := models.VideoItem{
					ID:        getStringFromMap(vm, "id"),
					Title:     getStringFromMap(vm, "title"),
					Thumbnail: getStringFromMap(vm, "thumbnail"),
					URL:       getStringFromMap(vm, "url"),
					Duration:  getStringFromMap(vm, "duration"),
				}
				videos = append(videos, video)
			}
		}
	}

	return videos
}

func getStringFromMap(m map[string]interface{}, key string) string {
	if v, ok := m[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

func getIntFromMap(m map[string]interface{}, key string, defaultVal int) int {
	if v, ok := m[key]; ok {
		switch val := v.(type) {
		case int:
			return val
		case float64:
			return int(val)
		}
	}
	return defaultVal
}

func downloadThumbnails(videos []models.VideoItem) {
	cacheService := services.GetVideoCacheService()
	for _, video := range videos {
		if video.Thumbnail != "" {
			cacheService.DownloadThumbnail(video.ID, video.Thumbnail)
		}
	}
}

func precacheVideos(videos []models.VideoItem) {
	cfg := config.Settings
	log.Printf("[预缓存] 开始预缓存 %d 个视频, 并发数: %d", len(videos), cfg.PrecacheConcurrent)
	sem := make(chan struct{}, cfg.PrecacheConcurrent)

	var wg sync.WaitGroup
	for _, video := range videos {
		wg.Add(1)
		go func(v models.VideoItem) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()
			precacheVideo(v.ID)
		}(video)
	}
	wg.Wait()
}

func precacheVideo(videoID string) {
	cacheService := services.GetVideoCacheService()
	scraperService := services.GetScraperService()
	proxyService := services.GetProxyService()
	cfg := config.Settings

	if cacheService.IsCached(videoID) {
		return
	}
	if cacheService.IsDownloading(videoID) {
		return
	}

	precacheQueue.RLock()
	if precacheQueue.set[videoID] {
		precacheQueue.RUnlock()
		return
	}
	precacheQueue.RUnlock()

	precacheQueue.Lock()
	precacheQueue.set[videoID] = true
	precacheQueue.Unlock()

	defer func() {
		precacheQueue.Lock()
		delete(precacheQueue.set, videoID)
		precacheQueue.Unlock()
	}()

	videoURL := cfg.TargetBaseURL + "/view_video.php?viewkey=" + videoID
	detail, err := scraperService.GetVideoDetailInNewTab(videoURL)

	if err != nil || detail == nil || detail.M3u8URL == "" {
		log.Printf("[预缓存] 跳过 %s: 无法获取视频链接", videoID)
		return
	}

	// 再次检查
	if cacheService.IsCached(videoID) || cacheService.IsDownloading(videoID) {
		return
	}

	videoSrc := detail.M3u8URL
	isMp4 := containsIgnoreCase(videoSrc, ".mp4") || !containsIgnoreCase(videoSrc, ".m3u8")

	if isMp4 {
		cacheService.StartMp4CacheDownload(videoID, videoSrc, detail)
	} else {
		// 获取m3u8内容
		client := proxyService.GetClient()
		resp, err := client.Get(videoSrc)
		if err != nil {
			return
		}
		defer resp.Body.Close()

		body := make([]byte, 1024*1024)
		n, _ := resp.Body.Read(body)
		originalM3u8 := string(body[:n])

		cacheService.StartCacheDownload(videoID, videoSrc, originalM3u8, detail)
	}

	log.Printf("[预缓存] 已启动: %s", videoID)
}

func containsIgnoreCase(s, substr string) bool {
	return strings.Contains(strings.ToLower(s), strings.ToLower(substr))
}
